crates/ingress-client/src/session.rs

```diff
@@ -0,0 +1,501 @@
+            partition_id = %self.partition,
+            "Starting ingress partition session",
+        );
+
+        loop {
+            state = match state {
+                SessionState::Connecting { retry } => self.connect(&ctx, retry).await,
+                SessionState::Connected { connection } => self.connected(&ctx, connection).await,
+                SessionState::Disconnected => SessionState::Connecting {
+                    retry: self.opts.connect_retry_policy.clone().into_iter(),
+                },
+                SessionState::Shutdown => {
+                    self.rx.close();
+                    break;
+                }
+            }
+        }
+    }
+
+    async fn connect(
+        &self,
+        ctx: &CancellationToken,
+        mut retry: RetryIter<'static>,
+    ) -> SessionState {
+        let Some(node_id) = self.partition_routing.get_node_by_partition(self.partition) else {
+            tokio::time::sleep(retry.next().unwrap_or_else(|| Duration::from_secs(1))).await;
+
+            return SessionState::Connecting { retry };
+        };
+
+        tokio::select! {
+            result = self
+                .networking
+                .get_connection(node_id, Swimlane::IngressData) => {
+                    match result {
+                        Ok(connection) => SessionState::Connected { connection },
+                        Err(ConnectError::Shutdown(_)) => SessionState::Shutdown,
+                        Err(err) => {
+                            debug!("Failed to connect to node {node_id}: {err}");
+                            tokio::time::sleep(retry.next().unwrap_or_else(|| Duration::from_secs(1))).await;
+                            SessionState::Connecting { retry }
+                        }
+                    }
+            }
+            _ = ctx.cancelled() => {
+                SessionState::Shutdown
+            }
+        }
+    }
+
+    /// Re-sends all inflight batches after a connection is restored.
+    async fn replay(&mut self, connection: &Connection) -> Result<(), ConnectionClosed> {
+        //todo(azmy): to avoid all the inflight batches again and waste traffic
+        //maybe test the connection first by sending an empty batch and wait for response
+        //before proceeding?
+
+        for batch in self.inflight.iter_mut() {
+            let Some(permit) = connection.reserve().await else {
+                return Err(ConnectionClosed);
+            };
+
+            // resend batch
+            let reply_rx = permit
+                .send_rpc(
+                    IngestRequest::from(Arc::clone(&batch.records)),
+                    Some(self.partition.into()),
+                )
+                .expect("encoding version to match");
+            batch.reply_rx = reply_rx.into();
+        }
+
+        Ok(())
+    }
+
+    async fn connected(&mut self, ctx: &CancellationToken, connection: Connection) -> SessionState {
+        if self.replay(&connection).await.is_err() {
```

Comment:
What happens if `replay` takes a long time and we cancel in the meantime?

Reply:
A answer could be that we want to finish the replay before reacting to the cancellation signal to make sure that we try sending all our batches to the receiver. But if this is needed, then it would probably be better to do this with `try_reserve` instead of `reserve().await`.